### ISSUE TITLE
CUDAContext::SwitchToDevice() acquires new CUDA stream instead of using the default stream

### ISSUE THREAD

**Issue by jwestell** [pytorch/pytorch#54830]
## 🐛 Bug

The first call to `caffe2::CUDAContext::SwitchToDevice(0)` will ignore the default CUDA Stream and acquire a new low priority CUDA Stream and set this stream as the new current stream for the given device. 

## To Reproduce

1. Run trace that allocates memory using the default stream and subsequently calls into `caffe2::CUDAContext::SwitchToDevice()`
2. Run the trace again any amount of times
3. Observe that the subsequent runs of the trace allocate memory from a stream that is not the default stream. CUDACachingAllocator is not able to re-use the memory allocated from the first execution because it was allocated from a separate stream.

Here is the portion of the stack trace in question:
```
caffe2::ThreadLocalCUDAObjects::GetCUDAStream(c10::StreamId stream_id, c10::DeviceIndex gpu, caffe2::ThreadLocalCUDAObjects * const this)
caffe2::ThreadLocalCUDAObjects::SetCurrentStreamId(c10::StreamId stream_id, c10::DeviceIndex gpu, caffe2::ThreadLocalCUDAObjects * const this) 
caffe2::CUDAContext::SwitchToDevice(c10::StreamId stream_id, caffe2::CUDAContext * const this) 
at::BaseContext::SwitchToDevice(caffe2::CUDAContext * const this)
```

## Expected behavior

Calls to `caffe2::CUDAContext::SwitchToDevice()` should use the default stream and not create a new stream. 

## Environment

```
PyTorch version: 1.7.0
Is debug build: False
CUDA used to build PyTorch: 10
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.6 (64-bit runtime)
Is CUDA available: N/A
CUDA runtime version: 11.1.74
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 455.23.04
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[conda] Could not collect

```

## Additional context

The constructor of `caffe2::ThreadLocalCUDAObjects` initializes a vector of CUDA stream for each GPU in the `cuda_streams_` member variable. These vectors are all initialized to empty and populated on subsequent calls to `caffe2::ThreadLocalCUDAObjects::GetCUDAStream`. One such code path is from the call stack shown above. Specifically, `caffe2::CUDAContext::SwitchToDevice(0)` is called from [context_base.h#L48](https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/caffe2/core/context_base.h#L48) (via [context_gpu.h#L180](https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/caffe2/core/context_gpu.h#L180)) with the apparent intention to use the default stream (based on notes in PR #15126). However, this code path triggers a call to `c10::cuda::getStreamFromPool` from within `caffe2::ThreadLocalCUDAObjects::GetCUDAStream` due to the fact that the `caffe2::ThreadLocalCUDAObjects::cuda_streams_` member has no reference to the default stream. This causes a new stream to be acquired (as per [tensor_cuda_stream](https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/docs/cpp/source/notes/tensor_cuda_stream.rst) documentation).

It seems that either the `cuda_streams_` member variable should be initialized to include the default stream, or `GetCUDAStream` should check for logical stream id 0 and call `getDefaultCUDAStream` instead of `getStreamFromPool`.

**Comment by jwestell** [comment]
Here is a unit test that demonstrates the issue:
```
TEST(CUDAContextTest, TestSwitchToDefaultDevice) {
  if (!HasCudaGPU()) return;

  at::cuda::CUDAStream defaultStream = at::cuda::getDefaultCUDAStream();
  at::cuda::CUDAStream currentStream = at::cuda::getCurrentCUDAStream();
  EXPECT_EQ(defaultStream, currentStream);

  CUDAContext context(0);
  context.SwitchToDevice();

  currentStream = at::cuda::getCurrentCUDAStream();
  EXPECT_EQ(defaultStream, currentStream);
}
```
The final expectation here fails:
```
168: ../caffe2/core/context_gpu_test.cc:27: Failure
168: Expected equality of these values:
168:   defaultStream
168:     Which is: stream 0 on device cuda:0
168:   currentStream
168:     Which is: stream 32 on device cuda:0
```

**Comment by jwestell** [comment]
@ezyang, what do you think of this issue? It seems that any custom operator derived from `caffe2::Operator<caffe2::CUDAContext>` will have `SwitchToDevice()` called in its constructor and subsequently create a new stream instead of using the default stream. Was this intended in #15174?

**Comment by ezyang** [comment]
It's been a long time, but I'm pretty sure this was intentional, because it's how Caffe2 used to work before I put the stream pools together. We had a big argument at the time about what the default stream behavior should be (use the default stream, or put it on some new "Caffe2 default stream but not actually CUDA default stream"); PyTorch's behavior won out but we never changed the Caffe2 semantics (and are very unlikely to do so).

**Comment by jwestell** [comment]
Thanks, @ezyang. I appreciate learning the history on that.

Here's the problem I'm facing related to this code path. I'm running detectron, and at some point during execution, the following stack trace is hit which constructs a `caffe2::Operator<caffe2::CUDAContext>` and subsequently calls `CUDAContext::SwitchToDevice()`. As described, this acquires a new CUDA stream and all subsequent memory allocations are associated with the new stream (all memory allocated prior to this from the CUDA default stream is not re-usable as per the CUDACachingAllocator logic for getting free blocks).

```
c10::cuda::setCurrentCUDAStream(c10::cuda::CUDAStream stream)
caffe2::ThreadLocalCUDAObjects::SetCurrentStreamId(0)
caffe2::CUDAContext::SwitchToDevice(0) 
at::BaseContext::SwitchToDevice(caffe2::CUDAContext * const this)
caffe2::Operator<caffe2::CUDAContext>::Operator()
caffe2::GenerateProposalsOp<caffe2::CUDAContext>::GenerateProposalsOp>()
caffe2::detail::_call_caffe2_op<caffe2::GenerateProposalsOp<caffe2::CUDAContext> >()
caffe2::detail::_call_caffe2_op_from_c10()
```

Can you suggest what the best approach is to work around this? Is it recommended to initialize my threads with the following  2 lines in order to acquire that initial CUDA stream for the caffe2 pool and get off the default CUDA stream early?
```
caffe2::CUDAContext context(0);
context.SwitchToDevice();
```
Thanks!

**Comment by ezyang** [comment]
Thank you for this context. And let me make sure I understand correctly: this is detectron from *PyTorch*, right? I think the right fix is to get detectron to stop using the Caffe2 ops, but maybe there is another fix that makes `call_caffe2_op_from_c10` properly use default stream.

**Comment by jwestell** [comment]
@ezyang, in factI am using [detectron2](https://github.com/facebookresearch/detectron2) which is exported with caffe2 tracing in order to use some caffe2 ops that are not available in the torchvision runtime. So, in general, I am running with a mix of pytorch and caffe2 ops. As soon as that first caffe2 op is constructed (as per the stack trace above), the CUDA stream is switched off of the default stream.

Is there a best practice to ensure the operators are all executed on the same stream?

**Comment by ezyang** [comment]
Nah, this is just a straight up bug. I haven't had a chance to investigate what a good way to solve it is.

**Comment by jwestell** [comment]
Fair enough, thanks for the response!

**Comment by malfet** [comment]
@jwestell when you talking about Detectron, are you referring to https://github.com/facebookresearch/Detectron or https://github.com/facebookresearch/detectron2 ?

**Comment by jwestell** [comment]
> @jwestell when you talking about Detectron, are you referring to https://github.com/facebookresearch/Detectron or https://github.com/facebookresearch/detectron2 ?

detectron2
